File: 0198-house-robber/0198-house-robber.cpp
vector<int> g;
const int MAX = 100 + 1;
int memory[MAX];
class Solution
{
public:
    int dp(int index)
    {
        if (index >= g.size())
            return 0;
        int &ret = memory[index];
        if (ret != -1)
            return ret;

        int leave = dp(index + 1);
        int take = g[index] + dp((index + 2));

        return ret = max(leave, take);
    }
    int rob(vector<int> &nums)
    {
        g = nums;
        memset(memory, -1, sizeof(memory));
        return dp(0);
    }
};

File: 0198-house-robber/0198-house-robber.java
import java.util.Arrays;

class Solution {
    int[] dp;

    public int rob(int[] nums) {
        dp = new int[nums.length];
        Arrays.fill(dp, -1);
        return solve(0, nums);

    }

    public int solve(int idx, int[] nums) {

        if (idx >= nums.length)
            return 0;
        
        if (dp[idx] != -1)
            return dp[idx];

        int skip = solve(idx + 1, nums);
        int take = nums[idx] + solve(idx + 2, nums);

        return dp[idx] = Math.max(skip, take);
    }
}